### PR TITLE
Create operator-dependencies-and-requirements.md

### PR DESCRIPTION
**Pull request by awgreene**

### CODE REVIEW

File: docs/operator-dependencies-and-requirements.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+An operator's dependencies are defined through the (Group, Version, Kind) of CustomResourceDefinitions (CRDs). An operator cannot define a dependency on a specific version (e.g. `etcd-operator v0.9.0`) or instance (e.g. `etcd v3.2.1`) of an operator. This encourages operator authors to depend on the interface and not the implementation; thereby allowing operators to update on individual release cycles.
+
+OLM manages operator dependency resolution very similarily to OS package managers, such as apt/dkpg and yum/rpm. However, given that operators are always running, OLM attempts to ensure that it never deploys a combination of operators that do not work with each other.
+
+This means that OLM will never:
+
+* install a set of operators that require APIs that can't be provided
+* update an operator in a way that breaks another that depends upon it
+
+### Dependency Resolution Alogorithm
```
Comment (ecordell) [inline docs/operator-dependencies-and-requirements.md]: I think that this is a bit too low-level. Can we describe how dependency resolution works without psuedocode like this?

Comment (awgreene) [reply docs/operator-dependencies-and-requirements.md]: Will do.

File: docs/operator-dependencies-and-requirements.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+Dependency resolution begins when a `subscription` is reconciled by OLM.
+
+When resolving a subscription, OLM will look at all subscriptions in the namespace and identify the latest offering of each operator. OLM will consider if any of these operators require an API that is missing from the cluster. For each of these missing APIs, OLM will select an operator that requires the missing API and prioritize its `CatalogSource` to search for the missing API. If the prioritized `CatalogSource` does not contain the API, OLM will search through the remaining `CatalogSources` in the namespace for the API. If an operator is found that provides the API, OLM will create a subscription for the operator using that `CatalogSource`.
```
Comment (ecordell) [inline docs/operator-dependencies-and-requirements.md]: > identify the latest offering of each operator

Might want to be careful about "latest" here - OLM will find the "next" update for the operator, which is controlled by the update graph (either an operator `replaces` or `skips` the current one, or there's an operator in the head of the channel with a `skipRange`)

File: docs/operator-dependencies-and-requirements.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+If the required API cannot be resolved OLM will not install operators that rely on that API.
+
+## Defining a Dependency
+
+An operator can define dependencies within its ClusterServiceVersion (CSV) by marking a CRD as `required`. A CRD can be marked as `required` by adding it to the `required array` within the `CustomResourceDefinitions` field, as shown in the example below:
```
Comment (ecordell) [inline docs/operator-dependencies-and-requirements.md]: Dependencies can also be specified via `spec.apiservicedefinitions.required`

Comment (awgreene) [reply docs/operator-dependencies-and-requirements.md]: Addressed in the latest version of the PR.

File: docs/operator-dependencies-and-requirements.md (hunk 2 above)
Comment (awgreene) [reply docs/operator-dependencies-and-requirements.md]: Addressed in the latest version of the PR.

File: docs/operator-dependencies-and-requirements.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+OLM manages operator dependency resolution very similarily to OS package managers, such as apt/dkpg and yum/rpm. However, given that operators are always running, OLM attempts to ensure that it never deploys a combination of operators that do not work with each other.
+
+This means that OLM will never:
+
+* install a set of operators that require APIs that can't be provided
+* update an operator in a way that breaks another that depends upon it
+
+### OLM Dependency Resolution
+
+Dependency resolution begins when a `subscription` is reconciled by OLM.
```
Comment (dinhxuanvu) [inline docs/operator-dependencies-and-requirements.md]: `Subscription` is an OLM object so it should be capitalized and put inside single quote just like `CatalogSource` and etc to keep the consistency across the doc.

File: docs/operator-dependencies-and-requirements.md [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+When resolving a subscription, OLM will look at all subscriptions in the namespace and identify the next offering of each operator based on the upgrade graph. By updating all subscriptions in the namespace, OLM avoids version deadlock that could be introduced when two operators have dependencies on each other.
```
Comment (dinhxuanvu) [inline docs/operator-dependencies-and-requirements.md]: Same above.

File: docs/operator-dependencies-and-requirements.md [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+Once OLM has identified the list of operator that should be installed in the namespace, OLM will attempt to resolve any required APIs that are missing from the cluster. For each of these missing APIs, OLM will select an operator that requires the missing API and prioritize its `CatalogSource` to search for the missing API. If the prioritized `CatalogSource` does not contain the API, OLM will search through the remaining `CatalogSources` in the namespace for the API. If an operator is found that provides the API, OLM will create a subscription for the operator using that `CatalogSource`.
+
+If the required API cannot be resolved OLM will not install operators that rely on that API.
+
+## Defining a Dependency
+
+An operator can define dependencies within its ClusterServiceVersion (CSV). An operator can specify:
```
Comment (dinhxuanvu) [inline docs/operator-dependencies-and-requirements.md]: `ClusterServiceVersion`

File: docs/operator-dependencies-and-requirements.md [hunk 7]
```diff
@@ -0,0 +1,40 @@
+# Operator Dependency and Requirement Resolution
+
+## Overview on Dependency Principles
+
+OLM manages the dependency resolution and upgrade lifecycle of running operators.
+
+An operator's dependencies are defined through the (Group, Version, Kind) of of an API. An operator cannot define a dependency on a specific version (e.g. `etcd-operator v0.9.0`) or instance (e.g. `etcd v3.2.1`) of an operator. This encourages operator authors to depend on the interface and not the implementation; thereby allowing operators to update on individual release cycles.
```
Comment (dinhxuanvu) [inline docs/operator-dependencies-and-requirements.md]: Double "of" in "(Group, Version, Kind) of of an API"

File: docs/operator-dependencies-and-requirements.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+
+### OLM Dependency Resolution
+
+Dependency resolution begins when a `subscription` is reconciled by OLM.
+
+When resolving a subscription, OLM will look at all subscriptions in the namespace and identify the next offering of each operator based on the upgrade graph. By updating all subscriptions in the namespace, OLM avoids version deadlock that could be introduced when two operators have dependencies on each other.
+
+Once OLM has identified the list of operator that should be installed in the namespace, OLM will attempt to resolve any required APIs that are missing from the cluster. For each of these missing APIs, OLM will select an operator that requires the missing API and prioritize its `CatalogSource` to search for the missing API. If the prioritized `CatalogSource` does not contain the API, OLM will search through the remaining `CatalogSources` in the namespace for the API. If an operator is found that provides the API, OLM will create a subscription for the operator using that `CatalogSource`.
+
+If the required API cannot be resolved OLM will not install operators that rely on that API.
```
Comment (dinhxuanvu) [inline docs/operator-dependencies-and-requirements.md]: Missing comma ",": `If the required API cannot be resolved, OLM will not install operators that rely on that API.`

File: docs/operator-dependencies-and-requirements.md (hunk 7 above)
Comment (awgreene) [reply docs/operator-dependencies-and-requirements.md]: Addressed.

File: docs/operator-dependencies-and-requirements.md (hunk 4 above)
Comment (awgreene) [reply docs/operator-dependencies-and-requirements.md]: Addressed in the latest version of the PR.

**Comment by awgreene** [comment]
Everything should be addressed @ecordell @dinhxuanvu

**Pull request merged**